crates/iota-genesis-builder/src/stake.rs

```diff
@@ -105,168 +103,304 @@ impl GenesisStake {
+                // Case 1. allocate gas stakes
+                if target_stake > 0 {
+                    // For staking gas coins we create a `TokenAllocation` object with
+                    // an empty `staked_with_timelock`
+                    self.create_token_allocation(delegator, target_stake, Some(validator), None);
+                }
+                // Case 2. allocate gas payments
+                if gas_to_pay > 0 {
+                    // For gas coins payments we create a `TokenAllocation` object with
+                    // `recipient_address` being the validator and no stake
+                    self.create_token_allocation(validator, gas_to_pay, None, None);
+                }
+            } else {
+                return Err(anyhow::anyhow!(
+                    "Not enough funds for delegator {:?}",
+                    delegator
+                ));
+            }
+        }
+
+        // If some surplus amount is left, then return it to the delegator
+        // In the case of a timelock object, it must be split during the `genesis` PTB
+        // execution
+        if let (Some(surplus_timelock), surplus_nanos) = timelock_surplus.take() {
+            self.timelocks_to_split
+                .push((surplus_timelock, surplus_nanos, delegator));
+        }
+        // In the case a the gas coin, it must be burned and the surplus re-allocated to
+        // the delegator (no split)
+        if let (Some(surplus_gas_coin), surplus_nanos) = gas_surplus.take() {
+            self.gas_coins_to_burn.push(surplus_gas_coin);
+            self.create_token_allocation(delegator, surplus_nanos, None, None);
+        }
+
+        Ok(())
+    }
 }
 
 /// The objects picked for token allocation during genesis
 #[derive(Default, Debug, Clone)]
-pub struct AllocationObjects {
-    inner: Vec<ObjectRef>,
+struct AllocationObjects {
+    /// The list of objects to burn for the allocations
+    to_burn: Vec<ObjectRef>,
     /// The total amount of nanos to be allocated from this
     /// collection of objects.
     amount_nanos: u64,
-    /// The surplus amount that is not be allocated from this
-    /// collection of objects.
-    surplus_nanos: u64,
+    /// The surplus object that should be split for this allocation. Only part
+    /// of its balance will be used for this collection of this
+    /// `AllocationObjects`, the surplus might be used later.
+    surplus: CoinSurplus,
     /// A (possible empty) vector of (amount, timelock_expiration) pairs
     /// indicating the amount to timelock stake and its expiration
     staked_with_timelock: Vec<(u64, u64)>,
 }
 
+/// The objects picked for token allocation during genesis
+#[derive(Default, Debug, Clone)]
+struct CoinSurplus {
+    // The reference of the coin to possibly split to get the surplus.
+    obj_ref: Option<ObjectRef>,
+    /// The surplus amount for that coin object.
+    surplus_nanos: u64,
+    /// Possibly indicate a timelock stake expiration.
+    timestamp: u64,
```

Comment:
Hmm.. I guess it makes sense to use an Option here if there should be a clear distinction between surplus being 0 and non-existent. In this case, I don't think it's necessary to convert this to an Option. It would add additional complexity to the code (checks).

Reply:
During the migration from UTXO to Object, [here](https://github.com/iotaledger/iota/blob/c42f61419852475adfbeb4eea8ab6aecf12f43aa/crates/iota-types/src/timelock/timelock.rs#L64) we check if the timelock still holds at the "target milestone timestamp". If it is [not timelocked at that time](https://github.com/iotaledger/iota/blob/c42f61419852475adfbeb4eea8ab6aecf12f43aa/crates/iota-genesis-builder/src/stardust/migration/migration.rs#L215) then we treat is as a simple basic output. This process is performed before this genesis building step, so we can't have unwanted behaviors like your scenario.